### PR TITLE
Single instance

### LINKED ISSUE #14: Support for single-window mode like KeePass
An option to force single window mode would be great. Currently, if keepass is opened while another instance is open there would be two instances. In single window mode it should just focus the currently open instance and not create another instance.

### LINKED ISSUE #99: Add ability to toggle window visibility (to/from Tray)
Hi everybody,

That's great we the users of KeePassX have a new dev team for the app!

It would be nice to have an Option called **Toggle Window Visibility** binded to a global key combination (like the auto-complete) to bring the main window to front and to minimize it with the same key combination. If the `minimize-to-tray` option is enabled, the minimize action of this option should minimize the window to tray.

_To give this issues a higher priority, I'm willing to contribute economically. Those who could implement it please specify how much should be collected and I could contribute to the implementation on [bountysource.com](https://www.bountysource.com/) or your crowdfunding site of preference._

### PR DESCRIPTION
**Pull request by antongulenko**
## Description
This PR merges [an older KeepassX PR](https://github.com/keepassx/keepassx/pull/171/files) that implements a single-instance mode, meaning that an existing keepassxc instance is raised instead of starting a second one. This has been proposed [here](https://github.com/keepassxreboot/keepassxc/issues/99#issuecomment-289321556), but I did not find a corresponding pull request.
For flexibility, a command line switch should probably be added to disable the single-instance mode if desired.

## Motivation and context
Single-instance mode allows raising keepassxc from the tray by running `keepassxc` from a script or a global keyboard shortcut.
Resolves #14.

## How has this been tested?
Compiled with the following cmake flags:
`-DWITH_TESTS=OFF -DWITH_XC_AUTOTYPE=ON -DWITH_XC_HTTP=ON -DWITH_ASAN=ON`
Everything seems to work fine. When running keepassxc repeatedly, it prints a message, notifies the running instance, and exits.

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/gui/Application.cpp [hunk 1]
```diff
@@ -85,6 +90,58 @@ Application::Application(int& argc, char** argv)
 #if defined(Q_OS_UNIX)
     registerUnixSignals();
 #endif
+
+    QString userName = qgetenv("USER");
+    if (userName.isEmpty()) {
+        userName = qgetenv("USERNAME");
+    }
+    QString identifier = "keepassx2";
```
Comment (TheZ3ro) [inline src/gui/Application.cpp]: This sould be changed to keepassxc

Comment (antongulenko) [inline src/gui/Application.cpp]: Commit pushed.

**Comment by droidmonkey** [comment]
I dont agree with this implementation as it adds thr networking dependency to all builds. We explicitly moved networking out of the core build so that a totally unconnected app can be built.

**Comment by phoerious** [comment]
Can't we do that via DBus instead?

**Comment by vsvyatski** [comment]
Just an idea: it can be theoretically done via a mutex. Some of you may not like Microsoft, but nevertheless I've found that idea on their site (it's MFC most likely, but the idea can be borrowed) https://support.microsoft.com/en-us/help/243953/how-to-limit-32-bit-applications-to-one-instance-in-visual-c
Qt has the [QMutex class](http://doc.qt.io/qt-5/qmutex.html).

**Comment by antongulenko** [comment]
Thanks for the feedback!
Unfortunately I lack the know-how to quickly re-implement this with another mechanism.
I hope that one of the devs will take this on, I think it would really increase the usability of keepassxc.

**Comment by TheZ3ro** [comment]
Better doing this with DBus or Mutex? @phoerious @droidmonkey

**Comment by droidmonkey** [comment]
Yes i think so. I do not want to reintroduce a network dependency for this feature. There are better ways to handle this.

**Comment by nicwortel** [comment]
I just wanted to point out that I stumbled upon this PR because I was looking for a solution to limit KeePassXC to a single instance. I often accidentally open a second instance instead of opening the existing one.

Since the linked issue (#99) is more about adding a key combination than about (the ability of) limiting to a single instance, I figured I'd drop this comment here.

As a final note, thanks everyone for your work! :+1:

**Comment by TheZ3ro** [comment]
Currently DBus is supported only on Linux

**Comment by sledgehammer999** [comment]
My 2 cents: There is also [qtsingleapplication](https://github.com/qtproject/qt-solutions/tree/master/qtsingleapplication). We use it in qBittorrent and seems to work across all platforms.

**Comment by TheZ3ro** [comment]
@sledgehammer999 thanks for the suggestion. 
I've seen that but unfortunately QtSimpleApplication is using QLocalServer and QLocalSocket.

From QLocalSocket documentation:
> On Windows this is a named pipe and on Unix this is a local domain socket.

But still you will need to compile with Qt5::Network and like @droidmonkey said:

> I do not agree with this implementation as it adds the networking dependency to all builds. We explicitly moved networking out of the core build so that a totally unconnected app can be built.

**Comment by Keisial** [comment]
What do you define as as «totally unconnected app»? And what is the goal of getting that?

You could call directly the low-level functions, thus avoiding the usage of Qt5::Network, but it would be doing the same in disguise. It may be worth if you just want to avoid linking with the library, though. It's not as if the lack of linking with Qt5::Network meant that the binary would be unable to perform network calls.

**Comment by droidmonkey** [comment]
Its more of a symbolic gesture to those individuals that want some greater assurance that their passwords are not potentially exposed to a network interface of any kind.

It seems to me that a temporary lock file could easily do the trick. Just have to be clever about app crashes (store pid in the file and check for running?)

**Comment by phoerious** [comment]
QLockFile does that already.

**Comment by TheZ3ro** [comment]
I was trying to develop a solution without using any Qt5::Network API.

The main thing is:
In the PR implementation the first started instance place a QLockFile. The newer instances will try to lock that file but seeing that is already locked will fail and connect to the QLocalServer socket. The first instance then will catch the connection, close it and raise the window.

I've tried replacing the QLocalServer with unix signals and (Win32 messages)[http://stackoverflow.com/a/140229] but that will make things more complicate.

Sincerely, I think we should include QLocalServer and use this PR. 
Otherwise, by using only QLockFile we cannot raise the first instance and only detect if KeePassXC is already running.

I agree that a password manager shouldn't use the network stack at all. But like specified in the Qt docs:

>   QLocalSocket: On Windows this is a named pipe and on Unix this is a local domain socket.

This is not "making network connection", it will works even without a networkadapter

**Comment by Keisial** [comment]
Not linking the Qt5::Network API is indeed symbolic. I would simply accept that library and replace it with an optional module that -on modern Linux- sets up a seccomp sandbox forbidding any socket(2) call where domain is not AF_LOCAL.
Done early enough, that guarantees the keepassxc program won't open a network connection (there are other means of exfiltration, of course). It should be admitted that KeePassHTTP would fail when such extension is enabled.

The main issue I see is that the seccomp filter is inherited through fork+execve. So, if the user opened an url, and that launched a new browser. That browser descendant of keepassxc would be unable to connect to the network (and probably fail bizarrely).

**Comment by droidmonkey** [comment]
I think this PR needs to be properly brought up to speed with the KeePassXC code base. We already integrate the Qt5::Network plugin with HTTP so it shouldn't be added to CMake...

**Comment by TheZ3ro** [comment]
@droidmonkey This is not meant to resolve #99 but rather #14

**Comment by droidmonkey** [comment]
Yah I know, i edited the description to say that :wink:

**Comment by TheZ3ro** [comment]
Missed the edit in the first post :smile:

**Comment by TheZ3ro** [comment]
What type of works this PR needs? Can we use QLocalServer for detecting running instances?

**Comment by droidmonkey** [comment]
Let's do it, makes the most sense

**Comment by TheZ3ro** [comment]
This PR already use it. It needs only a rebase and should be fine/ready to merge

**Comment by antongulenko** [comment]
I rebased the PR, please check if it matches your expectations.

**Comment by phoerious** [comment]
So, this still requires QNetworking?

**Comment by TheZ3ro** [comment]
@phoerious Yes, for QLocalSocket

File: src/main.cpp [hunk 2]
```diff
@@ -102,7 +107,15 @@ int main(int argc, char** argv)
...
+
+    QObject::connect(&app, &Application::anotherInstanceStarted,
+                    [&]() {
+                        mainWindow.ensurePolished();
+                        mainWindow.setWindowState(mainWindow.windowState() & ~Qt::WindowMinimized);
+                        mainWindow.show();
+                        mainWindow.raise();
+                        mainWindow.activateWindow();
+                    });
     QObject::connect(&app, SIGNAL(openFile(QString)), &mainWindow, SLOT(openDatabase(QString)));
```
Comment (phoerious) [inline src/main.cpp]: Shouldn't all the stuff after this be skipped when another instance is already running?

Comment (TheZ3ro) [inline src/main.cpp]: Actually there is a `return 0` at line 61 (if another instance is already running).

(Maybe should that be return 1 instead?)

**Pull request merged**